app/src/main/java/com/rkm/tasky/repository/implementation/TaskyEventRepositoryImpl.kt

```diff
@@ -0,0 +1,115 @@
+        withContext(dispatcher) {
+            return@withContext when (val results = localDataSource.getAllEventDetails(id)) {
+                null -> Result.Error(DatabaseError.ItemError.ITEM_DOES_NOT_EXIST)
+                else -> Result.Success(results.toEvent())
+            }
+        }
+
+    override suspend fun createEvent(event: Event): EmptyResult<NetworkError.APIError> =
+        withContext(dispatcher) {
+            val result = safeCall {
+                remoteDataSource.createEvent(
+                    event.asCreateEventRequest(),
+                    event.photosToUpload.asMultiPartBody()
+                )
+            }
+            result.onSuccess {
+                localDataSource.upsertEvent(it.asEventEntity())
+                attendeeLocalDataSource.upsertAttendees(it.attendees.map { attendee -> attendee.asAttendeeEntity() })
+                photoLocalDataSource.upsertPhotos(it.photos.map { photo -> photo.asPhotoEntity(it.id) })
+            }
+
+            result.onFailure {
+                syncDataSource.upsertSyncItem(
+                    SyncEntity(
+                        action = SyncUserAction.CREATE,
+                        item = SyncItemType.EVENT,
+                        itemId = event.id
+                    )
+                )
+            }
+
+            return@withContext result.asEmptyDataResult()
+        }
+
+    override suspend fun updateEvent(event: Event): EmptyResult<NetworkError.APIError> =
+        withContext(dispatcher) {
+            val result = safeCall {
+                remoteDataSource.updateEvent(
+                    event.asUpdateEventRequest(),
+                    event.photosToUpload.asMultiPartBody()
+                )
+            }
+            result.onSuccess {
+                localDataSource.upsertEvent(it.asEventEntity())
+                attendeeLocalDataSource.upsertAttendees(it.attendees.map { attendee -> attendee.asAttendeeEntity() })
+                photoLocalDataSource.upsertPhotos(it.photos.map { photo -> photo.asPhotoEntity(it.id) })
+            }
+
+            result.onFailure {
+                syncDataSource.upsertSyncItem(
+                    SyncEntity(
+                        action = SyncUserAction.UPDATE,
+                        item = SyncItemType.EVENT,
+                        itemId = event.id
+                    )
+                )
+            }
+
+            return@withContext result.asEmptyDataResult()
+        }
+
+    override suspend fun deleteEvent(event: Event): EmptyResult<NetworkError.APIError> = withContext(dispatcher) {
+        val result = safeCall { remoteDataSource.deleteEvent(event.id) }
+        result.onFailure {
+            SyncEntity(
+                action = SyncUserAction.DELETE,
+                item = SyncItemType.EVENT,
+                itemId = event.id
+            )
+        }
+        localDataSource.deleteEvent(event.asEventEntity())
```

Comment:
Two things:

1. I would first execute the `localDataSource` function and **only** when it succeeds, call the endpoint. 
2. Wrap the api call in `applicationScope.launch {}` to protect from cancellation when the user navigates back (and VM gets cleared, including the `viewModelScope` used to call this function)